### ISSUE TITLE
Knowledge Graph

### ISSUE THREAD

**Issue by AtesComp** [MISP/MISP#3571]
MISP seems so close to implementing true Knowledge Graph technology as provided through the Semantic Web community. See the Google Knowledge Graph, Facebook Graph, and DBpedia as examples. MISP can help by pushing the OASIS CTI to develop a true ontology for STIX, CyBOX, MAEC, etc. The ontology principles, graph DBs, and inference engines seem to be a perfect match for MISP. MISP Galaxies could be federated Knowledge Graphs / SPARQL Endpoints. The taxonomies need to move to the next level--employing actual ontologies. What needs to be done to move MISP in this direction?

**Comment by iglocska** [comment]
I think the nr1 thing we need before any such move is even considered is real world benefits to our day to day work. Generally our design process is:

1. Real-world need to solve a problem pops-up
2. We try to come with the least complex solution to the above issue

This avoids over-engineering for something that results in little in terms of day-to-day benefits.

**Comment by adulau** [comment]
Thank you very much for your feedback.  

We accept pull-request with improved software implementation for such kind of new data models. We would be interested to get practical implementation especially to evaluate the benefit over simple and existing formats.

Feel free to reopen the issue with the associated pull-requests.

Thank you.

**Comment by AtesComp** [comment]
Thank you for responding.  I hope to clarify my comment with a more complete discussion below.  I will attempt to answer Andras' concerns.

I'm suggesting that MISP is replicating in a specific instance exactly what the Semantic Web model and the Resource Description Framework (RDF) proposes.  MISP objects are connected to attributes and other objects--a data graph.  All elements are equivalent to typed nodes.  The links are predicated edges.  Even MISP tags are simply a bag of enrichment nodes connected to an object. So, the MISP storage model is well suited to a simpler graph database.  There are many suitable RDF graph databases.  See MarkLogic (https://www.marklogic.com/product/marklogic-database-overview/) for one example.  Graph DBs are making important advancements in technology (https://www.datastax.com/2017/12/another-graph-database-announcement-another-shout-out-for-datastax-funded-innovation).  Layering technology is also available to transform traditional RDBMS systems into graph DBs. See D2RQ (http://d2rq.org/) for an example.

I make the following statements without qualification.  Among others, Semantic Web solutions provide for the following salient points:
- Provides for interchangeable format systems to exchange data removing the need for rewrites when information changes
- Technology utilizes flexible storage—the storage schema doesn’t require change as problem complexity increases
- Next generation of visualization through browsers
- Ability to add analytical data through reasoning

Semantic Web technology simplifies the storage layer removing much (if not all) of the traditional RDBMS database management.  Ontology information is stored in the same manner as the working data.  The traditional RDBMS storage layer requires schema modification when requirements change and places a higher burden on maintenance and development.  To modify an RDF Graph DB, the ontology is modified--a simple matter of modifying ontology data (triples) in the store.

Comparison:
- Traditional RDBMSs are suited for well defined, static applications
- Graph DBs are suited for changing, dynamic applications
- They are both RESTful.

I pose that the IOC collection and sharing domain is a highly dynamic domain, particularly when analysis is required.

To extract data for use, the Semantic Web model is a simpler method for extraction, display, and manipulation.  Most SPARQL endpoints have built in methods for transformation into HTML, JSON, XML, RDF/XML, CSV, etc.  See https://dbpedia.org/sparql for an example.  Even queries directly transformed to Google Visualization JSON have been done.  I don't suggest providing a raw SPARQL endpoint for user interaction (it could be an option), but point out that transforms are natural extensions at query time and can be built as needed.  NIDS signature exports on a specific query mapping could be done in this manner.  This would easily allow for custom exports to be applied.

The more important advantage for using this technology is the built-in machine reasoning.  Whereas MISP builds a "correlation" system to link related events, Semantic Web technology has correlation as a "built-in" solution.  Semantic Web ontologies provide inference relations: class-subclass, property-subproperty, transitive, asymmetric, symmetric, functional, inverse functional, and much more.  Class restrictions further enhance the inference capability.  As defined by the Web Ontology Language (OWL2), the capabilities are available in the technology.  As a correlating example, a SPARQL query would be constructed to ask: "query for all events that have the same correlating feature or features where some features may not be the same or even present, rank them on equality".  This type of query is a primary reason the technology was developed.  The Semantic Web solution is designed to provide powerful reasoning with minimal data.

Federation is built into SPARQL and doesn't break when a federated query endpoint is not available.  It designed for open exchange.  The Galaxy solution just reinvented this idea already present in Semantic Web tech.

In addition, IOC collections are moving to Big Data analytics.  Using this model, allows collections to add feature sets or transform existing correlations to feature set for other machine learning processes.  The additions/modification do not require any additional DB modification.  Just add ontology statements to the dataset.

The documented "Populate for OpenIOC" makes another strong point for using Semantic Web tech.  While some indicators may not fit the MISP model, when using an graph DB, you can store and link the unmatched indicators anyway.  When the ontology is updated to employ the indicators, the system would  use them natively.  Until then, they could still be stored and queried for other user purposes.  Semantic Web tech doesn't fail just because we don't have current definitions for the data.  In fact, we would just import an OpenIOC ontology to use their definitions directly, if it were available.

I haven't event discussed the issues related to classification and SCI controls.  I see none of this in MISP.  Using a Semantic Web model would make this quite a bit easier with an ontology as opposed to building specific control software to manage access, dissemination, redaction.

**As a simple example**, at the DBpedia SPARQL endpoint (https://dbpedia.org/sparql), use the provided query "select distinct ?Concept where {[] a ?Concept} LIMIT 100".  This displays the first 100 "typed" blank nodes.  The types are many.  One is "http://xmlns.com/foaf/0.1/Person". Another is "http://dbpedia.org/ontology/Person".  This indicated two competing/similar/different class definitions of a person. Given that all "people" in the DBpedia are typed as "http://dbpedia.org/ontology/Person" and "http://dbpedia.org/ontology/Person" is a subclass of or equivalent to "http://xmlns.com/foaf/0.1/Person", we should be able to query for all entities that are of the general type "http://xmlns.com/foaf/0.1/Person".  Use the query "select ?s where { ?s a \<http://xmlns.com/foaf/0.1/Person> } LIMIT 100".  Since nothing in DBpedia was directly typed as the FOAF person type, the query infers the class relation to the DBpedia person types.  In fact, we can see all ontological relations for "http://dbpedia.org/ontology/Person" by querying "select ?p ?o where { \<http://dbpedia.org/ontology/Person> ?p ?o } LIMIT 100".  As all results are in HTML format, you can drill down directly on the links to examine a result element.

I'm suggesting MISP development has naturally gone in the direction of what  developers know (a "go with what you know" principle) and not what might be the best, simplest technology.  That employing a knowledge graph with Semantic Web technology is a least complex solution that provides a better knowledge research platform.  This is precisely why Big Data analytics, such as Google, Facebook, Walmart, BBC, and any number of other organizations use this as their base technology.  Traditional vs. Semantic Web tech is akin to comparing trigonometry to vector calculus.  Trig will get you the answer for the intersecting point of 3 planes, but vector calculus is way easier once you understand it.

I hope this clarifies the issues by providing several demonstrative use cases.  I wouldn't characterize this as over-engineering a solution.  Instead, I see it as understanding and fitting the best technology to the requirements.  I would like to continue this as a lively discussion to benefit this platform.  It's simply hard to move forward without discussing the use of a MISP/STIX/IOC ontology.  However, the definitions in the taxonomies can get us there.

**Comment by adulau** [comment]
Thank you for your input but it's quite dense and going in various directions. To keep the discussion interesting in the scope of MISP development. I would like to refine it on concrete example.

Can you model a specific use-case (with a simple data model example using RDF) to represent a forensic analysis case or evidence acquisition in RDF based on your experience in the field?

**Comment by AtesComp** [comment]
Sorry, its been awhile since I've been able to get back to this.

The input was dense, but it is only going in one direction--employ the back end data storage as a native knowledge graph system.  The use case is to implement a system that simplifies the storage management and investigative process.

The current system using an RDBMS requires schema management to modify structure--tables, fields, etc.--to implement new features.  Then, any application parts dependent on the changed tables may break due to changes in the schema.

For a semantic web based system, the data and its related structure is simply added to the graph as triples.  The current application doesn't "break" due to these additions.  Updates on the application can incorporate the new data and "schema" as needed.

Simple Use Case:
I have some data,
Actor1 --- rdf:type --> ont:MalwareDistributor
Actor1 --- ont:gender --> ont:Male
Actor1 --- ont:visitsSuspectSite --> ont:SiteX
ont:SiteX --> ont:hasIP --> "123.456.789.123"

I find more data,
Actor2 --- rdf:type --> foaf:Person
Actor3 --- rdf:type --> foaf:Person
Actor3 --- ont:hasSon --> Actor2
Actor3 --- ont:gender --> ont:Female

Now I create some rules (they are just more data elements):
ont:MalwareDistributor --- rdf:subProperty --> foaf:Person
ont:hasSon --- rdf:subProperty --> ont:hasChild
ont:hasChild --- rdf:subProperty --> ont:hasFamilyRelation
ont:hasMother --- rdf:subProperty --> ont:hasParent
ont:hasParent --- rdf:subProperty --> ont:hasFamilyRelation
ont:hasFamilyRelation --- rdf:subProperty --> foaf:knows
ont:hasParent -- owl:inverseOf --> ont:hasChild

I can also create triples that represent the rule X --- ont:hasMother --> Y if Y --- ont:gender --> ont:Female AND Y --- ont:hasChild --> X

...and others...

With a reasoner engine native to the knowledge graph store, I can now query "Does Actor2 have a mother?"
The data does not explicitly say Actor2 has a mother, but the reasoner reports that Actor2 does indeed have a mother--Actor3.

Now, I find new data:
Actor2 --- ont:hasDegree --> ont:BSCS
Actor2 --- ont:visitsSuspectSite --> ont:SiteX
...and others...

With the known data (even with timestamps), we can't determine a positive relation between Actor1 and Actor2

Later we find:
Actor1 --- ont:hasSibling --> Actor3
This implies there is a relationship between Actor3 and Actor1--and uncle relationship.  Added rules expose the relationship without needing to modify the application.
X --- ont:hasFamilyRelation --> Z if X --- ont:hasParent --> Y AND Y --- ont:hasSibling --> Z
X --- ont:hasUncle --> Z if X --- ont:hasParent --> Y AND Y --- ont:hasSibling --> Z AND Z --- ont:gender --> ont:Male
ont:hasUncle --- rdf:subProperty --> ont:hasFamilyRelation
...and others...

We query "Who knows Actor1 and visits the same suspect site?"  Actor2 is the answer.  Another query identifies how Actor1 knows Actor2--the uncle relationship.

Eventually, we positively confirm Actor2 is the malware author and Actor1 is receiving update via SiteX.

While wildly oversimplified, this captures some of the issues.  In this model, we don't rely on the table structures having proper indices linking the records...it's a consequence of having a proper knowledge graph linking the data naturally.  I can add the data without worrying about its structure and immediate use.  I can develop general queries and add them to the application at will.  Graph displays for adhoc drilling are a natural extension.
